### PR TITLE
Enhance quantization api

### PR DESCRIPTION
**Pull request by xinyu-intel**
## Description ##
Enhance quantization api to make it more general for users to integrate.

Motivation is that original quantized_model api only supports image classification workloads which have only one input and only support DataIter as input. It's more reasonable to move mod related parts outside the api for users to customize.

So, this PR split original api into three part:

1. quantize_graph: quantize fp32 model to int8 model w/o calibration and return a collector for collecting calibration information in the next step.

2. [outside api]: users need only add a few lines together with mod.forward for collecting calibration information.

3. calib_graph: generate calibrated model based on filled collector.

Another one is ```quantize_net```, which is for quantizing a gluon hybridblock and load it back to gluon as a symbolblock.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: example/quantization/imagenet_gen_qsym_mkldnn.py [hunk 1]
```diff
@@ -279,12 +279,17 @@ def save_params(fname, arg_params, aux_params, logger=None):
...
+                                                              excluded_sym_names=excluded_sym_names,
+                                                              calib_mode=calib_mode, calib_layer=calib_layer,
+                                                              quantized_dtype=args.quantized_dtype, logger=logger)
     if calib_mode == 'none':
-        logger.info('Quantizing FP32 model %s' % args.model)
-        qsym, qarg_params, aux_params = quantize_model(sym=sym, arg_params=arg_params, aux_params=aux_params,
-                                                       ctx=ctx, excluded_sym_names=excluded_sym_names,
-                                                       calib_mode=calib_mode, quantized_dtype=args.quantized_dtype,
-                                                       logger=logger)
+        # logger.info('Quantizing FP32 model %s' % args.model)
```
Comment (ciyongch) [inline example/quantization/imagenet_gen_qsym_mkldnn.py]: remove commented/useless codes.

**Review by ciyongch — COMMENTED** [review]
There's many duplicated codes in quantization.py, can we use `quantize_graph` and `calib_graph` to construct `quantize_symbol` for legacy usage?
BTW, any concrete example to use such feature is helpful the motivation :)

File: python/mxnet/contrib/quantization.py [hunk 2]
```diff
@@ -543,3 +546,249 @@ def quantize_model(sym, arg_params, aux_params,
...
+    import tempfile
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        prefix = os.path.join(tmpdirname, 'tmp')
+        network.export(prefix, epoch=0)
+        symnet, args, auxs = mx.model.load_checkpoint(prefix, 0)
+    logger.info('Exclude all FullyConnect and Flatten layers.')
+    if exclude_layers is None:
+        exclude_layers = []
+    for layers in list(symnet.get_internals()):
+        if layers.name.find('dense') != -1 or layers.name.find('flatten') != -1:
```
Comment (ciyongch) [inline python/mxnet/contrib/quantization.py]: Some FC layers might have other names like "fc1/fc2..."?

File: python/mxnet/contrib/quantization.py [hunk 3]
```diff
@@ -543,3 +546,249 @@ def quantize_model(sym, arg_params, aux_params,
...
+            exclude_layers.append(layers.name)
+    symnet = symnet.get_backend_symbol('MKLDNN')
```
Comment (ciyongch) [inline python/mxnet/contrib/quantization.py]: Please use MKLDNN_QUANTIZE here too, to avoid some fusion that only appropriate to FP32 flow.

File: python/mxnet/contrib/quantization.py (hunk 2 above)
Comment (xinyu-intel) [reply python/mxnet/contrib/quantization.py]: yes, add it and it's a short-term trick here. For gluonnlp, there are more sophisticated names...support loading int8 weight in gluon is necessary.

File: python/mxnet/contrib/quantization.py (hunk 3 above)
Comment (xinyu-intel) [reply python/mxnet/contrib/quantization.py]: done

File: example/quantization/imagenet_gen_qsym_mkldnn.py (hunk 1 above)
Comment (xinyu-intel) [reply example/quantization/imagenet_gen_qsym_mkldnn.py]: done
